src/nm-checkpoint.c

```diff
@@ -647,6 +725,10 @@ dispose (GObject *object)
 
 	g_clear_pointer (&priv->devices, g_hash_table_unref);
 	g_clear_pointer (&priv->connection_uuids, g_hash_table_unref);
+	if (priv->removed_devices)
+		g_ptr_array_unref (priv->removed_devices);
+
+	nm_clear_g_signal_handler (priv->manager, &priv->dev_removed_id);
```

Comment:
`priv->manager` does not own a reference to NMManager.  Previously, that was less of a problem, because every use of `priv->manager` happened at points where the manager clearly was still alive (e.g. during `nm_checkpoint_rollback()`).

It's not clear that at the moment when the last reference to NMCheckpoint is dropped, that the manger is still alive. You need to add a reference, and `g_object_clear()` it in dispose.

Note that then NMManager and NMCheckpoint hold a reference to each other. But that is fine, because during shutdown, NMManager must first destroy all checkpoints (giving up it's reference and break the cycle).

Of course, currently we just utterly leak NMManager and it wouldn't matter. But if we ever hope to fix that, all places must get this right (first).

Reply:
But shouldn't we expect to be the manager the last one to be finalized (the scenario is NM shutting down, right)?
Anyway, agreed, linking an object without taking a reference is ugly.
Added a new commit on top of the branch for this.